lib/Runtime/Language/JavascriptOperators.cpp

```diff
@@ -1357,17 +1357,50 @@ namespace Js
 
         RecyclableObject* instance = RecyclableObject::FromVar(instanceVar);
         ScriptContext* scriptContext = instance->GetScriptContext();
+
+        if (!PHASE_OFF1(IsConcatSpreadableCachePhase))
+        {
+            BOOL retVal = FALSE;
+            Type *instanceType = instance->GetType();
+            IsConcatSpreadableCache *isConcatSpreadableCache = scriptContext->GetThreadContext()->GetIsConcatSpreadableCache();
+
+            if (isConcatSpreadableCache->TryGetIsConcatSpreadable(instanceType, &retVal))
+            {
+#ifdef ENABLE_DEBUG_CONFIG_OPTIONS
+                if (Js::Configuration::Global.flags.Trace.IsEnabled(Phase::IsConcatSpreadableCachePhase))
+                {
+                    Output::Print(_u("IsConcatSpreadableCache hit: %p\n"), instanceType);
+                }
+#endif
+                return retVal;
+            }
+
+            Var spreadable = JavascriptOperators::GetProperty(instance, PropertyIds::_symbolIsConcatSpreadable, scriptContext);
+
+            if (spreadable != scriptContext->GetLibrary()->GetUndefined())
```

Comment:
Thanks. That seems odd. It doesn’t distinguish between a non-existent property and a property whose value is “undefined”. But probably not worth raising as a spec issue.

--Paul

From: suwc [mailto:notifications@github.com]
Sent: Friday, June 24, 2016 11:48 AM
To: Microsoft/ChakraCore <ChakraCore@noreply.github.com>
Cc: Paul Leathers <pleath@microsoft.com>; Team mention <team_mention@noreply.github.com>
Subject: Re: [Microsoft/ChakraCore] Enable ES6IsConcatSpreadable under experimental (#1195)


In lib/Runtime/Language/JavascriptOperators.cpp<https://github.com/Microsoft/ChakraCore/pull/1195#discussion_r68444357>:

> +            IsConcatSpreadableCache *isConcatSpreadableCache = scriptContext->GetThreadContext()->GetIsConcatSpreadableCache();

> +

> +            if (isConcatSpreadableCache->TryGetIsConcatSpreadable(instanceType, &retVal))

> +            {

> +#ifdef ENABLE_DEBUG_CONFIG_OPTIONS

> +                if (Js::Configuration::Global.flags.Trace.IsEnabled(Phase::IsConcatSpreadableCachePhase))

> +                {

> +                    Output::Print(_u("IsConcatSpreadableCache hit: %p\n"), instanceType);

> +                }

> +#endif

> +                return retVal;

> +            }

> +

> +            Var spreadable = JavascriptOperators::GetProperty(instance, PropertyIds::_symbolIsConcatSpreadable, scriptContext);

> +

> +            if (spreadable != scriptContext->GetLibrary()->GetUndefined())

Yes, the spec (22.1.3.1.1) says:

3.If spreadable is not undefined, return ToBoolean(spreadable).
4.Return ? IsArray(O).

________________________________

In reply to: 68430889<https://github.com/Microsoft/ChakraCore/pull/1195#discussion_r68430889>

—
You are receiving this because you are on a team that was mentioned.
Reply to this email directly, view it on GitHub<https://github.com/Microsoft/ChakraCore/pull/1195/files/f75710dc77f9f83fa1a71999856918a55e1007ec#r68444357>, or mute the thread<https://github.com/notifications/unsubscribe/APF8RD5rnqA-YfYm6xUWlKapt1-vW5d1ks5qPCZUgaJpZM4I9cJU>.